README.md
# shinyutils
Various utilities for common tasks. :sparkles: :sparkles: :sparkles:

## Setup
Install with `pip`. Additional features can be enabled with the `[<feature>]` syntax shown below. Available optional features are:

* `color`: color support for logging

* `plotting`: support for `matplotlib` and `seaborn`

```bash
pip install shinyutils  # basic install
pip install "shinyutils[color]"  # install with color support
pip install "shinyutils[color,plotting]"  # install with color and plotting support
```

## Components

### `subcls`
Utility functions for dealing with subclasses.

#### Functions

* __`get_subclasses(cls)`__: returns a list of all the subclasses of `cls`.

* __`get_subclass_names(cls)`__: returns a list of names of all subclasses of `cls`.

* __`get_subclass_from_name(base_cls, cls_name)`__: return the subclass of `base_cls` named `cls_name`.

### `argp`
Utilities for argument parsing.

#### `CommaSeparatedInts`
`ArgumentParser` type representing a list of `int` values. Accepts a string of comma separated values, e.g., `'1,2,3'`.

#### `InputFileType`
`FileType` restricted to input files, (with `'-'` for `stdin`). Returns a `file` object.

#### `OutputFileType`
`FileType` restricted to output files (with `'-'` for `stdout`). The file's parent directories are created if needed. Returns a `file` object.

#### `InputDirectoryType`
`ArgumentParser` type representing a directory. Returns a `Path` object.

#### `OutputDirectoryType`
`ArgumentParser` type representing an output directory. The directory is created if it doesn't exist. Returns a `Path` object.

#### `ClassType`
`ArgumentParser` type representing sub-classes of a given base class. The returned value is a `class`.

```python
class Base:
    pass

class A(Base):
    pass

class B(Base):
    pass

arg_parser.add_argument("--cls", type=ClassType(Base), default=A)
```

#### `KeyValuePairsType`
`ArgumentParser` type representing mappings. Accepts inputs of the form `str=val,[...]` where val is `int/float/str`. Returns a `dict`.

### `logng`
Utilities for logging.

#### `build_log_argp`
Creates an argument group with logging arguments.

```python
>>> arg_parser = ArgumentParser()
>>> _ = build_log_argp(arg_parser)  # returns the parser
>>> arg_parser.print_help()
usage: -c [-h] [--log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}]

optional arguments:
  -h, --help            show this help message and exit
  --log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}
```

#### `conf_logging`
Configures global logging using arguments returned by `ArgumentParser.parse_args`. `log_level` can be over-ridden with the keyword argument. Colors (enabled by default if `rich` is installed) can be toggled.

```python
args = arg_parser.parse_args()
conf_logging(args)
conf_logging(args, log_level="INFO")  # override `log_level`
conf_logging(use_colors=False)  # disable colors
```

When imported, `shinyutils` calls `conf_logging` without any arguments.

### `matwrap`
Wrapper around `matplotlib` and `seaborn`.

#### `MatWrap`

```python
from shinyutils.matwrap import MatWrap as mw  # do not import `matplotlib`, `seaborn`

mw.configure()  # this should be called before importing any packages that import matplotlib

fig = mw.plt().figure()
ax = fig.add_subplot(111)  # `ax` can be used normally now

# Use class methods in `MatWrap` to access `matplotlib`/`seaborn` functions.
mw.mpl()  # returns `matplotlib` module
mw.plt()  # returns `matplotlib.pyplot` module
mw.sns()  # returns `seaborn` module
```

Use `mw.configure` to configure plots. Arguments (defaults in bold) are:

* `context`: seaborn context (__paper__/poster/talk/notebook)

* `style`: seaborn style (white/whitegrid/dark/darkgrid/__ticks__)

* `font`: any font available to fontspec (default __Latin Modern Roman__)

* `latex_pkgs`: additional latex packages to be included before defaults

* `**rc_extra`: matplotlib rc parameters to override defaults

`mw.configure()` is called when `shinyutils.matwrap` is imported.

Use `add_parser_config_args` to add matwrap config options to an argument parser.

```python
>>> arg_parser = ArgumentParser()
>>> _ = mw.add_parser_config_args(arg_parser, group_title="plotting options")  # returns the parser group
>>> arg_parser.print_help()
usage: -c [-h] [--plotting-context {paper,notebook,talk,poster}]
          [--plotting-style {white,dark,whitegrid,darkgrid,ticks}]
          [--plotting-font PLOTTING_FONT]
          [--plotting-latex-pkgs PLOTTING_LATEX_PKGS [PLOTTING_LATEX_PKGS ...]]
          [--plotting-rc-extra PLOTTING_RC_EXTRA]

optional arguments:
  -h, --help            show this help message and exit

plotting options:
  --plotting-context {paper,notebook,talk,poster}
  --plotting-style {white,dark,whitegrid,darkgrid,ticks}
  --plotting-font PLOTTING_FONT
  --plotting-latex-pkgs PLOTTING_LATEX_PKGS [PLOTTING_LATEX_PKGS ...]
  --plotting-rc-extra PLOTTING_RC_EXTRA
```

`group_title` is optional, and if omitted, matwrap options will not be put in a separate group.

#### Plot
`Plot` is a wrapper around a single matplotlib plot, designed to be used as a context manager.

```python
from shinyutils.matwrap import Plot

with Plot(save_file, title, sizexy, labelxy, logxy) as ax:
  ...
```

Only the `save_file` argument is mandatory. When entering the context, `Plot` returns the plot axes, and when leaving, the plot is saved to the provided path.

### `pt`
Utilities for pytorch.

#### `PTOpt`
Wrapper around pytorch optimizer and learning rate scheduler.

```python
from shinyutils.pt import PTOpt
opt = PTOpt(
        weights,  # iterable of parameters to update
        optim_cls,  # subclass of torch.optim.Optimizer
        optim_params,  # arguments to initialize optim_cls
        lr_sched_cls,  # subclass of torch.optim.lr_scheduler._LRScheduler
        lr_sched_params,
)
...
opt.zero_grad()
loss.backward()
opt.step()
```

`lr_sched_` arguments are optional, and control the learning rate schedule. The class can also be used with argument parsers.

```python
>>> arg_parser = ArgumentParser(formatter_class=LazyHelpFormatter)
>>> PTOpt.add_parser_args(
        arg_parser,
        arg_prefix="test",  # all options will be prefixed with "test-"
        group_title="pt test",  # if None, separate group will not be created
        default_optim_cls=Adam,
        default_optim_params=None,  # if None, default is an empty dict
        add_lr_decay=True,
    )
>>> arg_parser.print_help()
options:
  -h, --help                            show this help message and exit (optional)

pt test:
  --test-optim-cls cls                  ({Adadelta / Adagrad / Adam / AdamW / SparseAdam /
                                          Adamax / ASGD / SGD / Rprop / RMSprop / LBFGS}
                                          default: Adam)
  --test-optim-params str=val[,...]     (default: {})
  --test-lr-sched-cls cls               ({LambdaLR / MultiplicativeLR / StepLR / MultiStepLR /
                                        ExponentialLR / CosineAnnealingLR / CyclicLR /
                                        CosineAnnealingWarmRestarts / OneCycleLR} optional)
  --test-lr-sched-params str=val[,...]  (default: {})
>>> args = arg_parser.parse_args(...)
>>> opt = PTOpt.from_args(weights, args, arg_prefix="test")
```

`PTOpt` can also add help options to argument parsers to display signatures for optimizer and learning rate schedule classes.

```python
>>> arg_parser = ArgumentParser()
>>> PTOpt.add_help(arg_parser)
>>> arg_parser.print_help()
usage: -c [-h] [--explain-optimizer EXPLAIN_OPTIMIZER]
          [--explain-lr-sched EXPLAIN_LR_SCHED]

optional arguments:
  -h, --help            show this help message and exit

pytorch help:
  --explain-optimizer EXPLAIN_OPTIMIZER
                        describe arguments of a torch optimizer
  --explain-lr-sched EXPLAIN_LR_SCHED
                        describe arguments of a torch lr scheduler

>>> arg_parser.parse_args(["--explain-optimizer", "Adam"])
Adam(params, lr=0.001, betas=(0.9, 0.999), eps=1e-08, weight_decay=0, amsgrad=False)
...
```

#### `FCNet`
`FCNet` is a template class for fully connected networks.

```python
from shinyutils.pt import FCNet
net = FCNet(
        in_dim,  # input dimension
        out_dim,  # output dimension
        hidden_sizes,  # list of hidden layer sizes
        hidden_act,  # hidden layer activations (default relu)
        out_act,  # output layer activation (default None)
)
```

Like `PTOpt`, this class also supports construction through command line arguments.

```python
>>> arg_parser = ArgumentParser(formatter_class=LazyHelpFormatter)
>>> FCNet.add_parser_args(
        arg_parser,
        arg_prefix="test",
        group_title="fcnet",
        default_indim=None,  # None means the option is mandatory
        default_outdim=1,
        default_hidden_sizes=None,
        default_hidden_act=F.relu,
        default_out_act=None,  # here, None means no output activation
)
>>> arg_parser.print_help()
options:
  -h, --help                           show this help message and exit (optional)

fcnet:
  --test-fcnet-indim int               (required)
  --test-fcnet-outdim int              (default: 1)
  --test-fcnet-hidden-sizes int,[...]  (required)
  --test-fcnet-hidden-act func         (default: relu)
  --test-fcnet-out-act func            (optional)
>>> args = arg_parser.parse_args(...)
>>> net = FCNet.from_args(args, arg_prefix="test")
```

#### `NNTrainer`
This class trains a model on a dataset, and accepts multiple dataset "formats".

```python
from shinyutils.pt import *
nn_trainer = NNTrainer(
    batch_size,  # only mandatory argument
    data_load_workers,  # default 0
    shuffle,  # default True
    pin_memory,  # default True
    drop_last,  # default True
    device,  # default cuda if available else cpu
)
nn_trainer.set_dataset(
    dataset,  # can be a torch Dataset, a tuple of torch Tensors, or a tuple of numpy arrays
)
model = FCNet(...)
opt = PTOpt(...)
loss_fn = torch.nn.functional.mse_loss
nn_trainer.train(model, opt, loss_fn, iters)
```

#### `SetTBWriterAction`
`argparse` action to create a tensorboard summary writer. The writer is stored in the `tb_writer` attribute of the argument namespace; this can be overridden by setting `SetTBWriterAction.attr`. The usage is shown below.

```python
parser.add_argument(
    "--tb-dir",
    type=OutputDirectoryType(),
    help="tensorboard log directory",
    default=None,
    action=SetTBWriterAction,
)
parser.set_defaults(**{SetTBWriterAction.attr: Mock(SummaryWriter)})
```

`parser.tb_writer` will contain a `SummaryWriter` like object. If no log directory is provided through the command line, this object will be a dummy. So tensorboard functions can be called on the writer without extra checks.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [12.3.0](https://github.com/jayanthkoushik/shinyutils/compare/v12.2.1...v12.3.0) (2021-12-09)


### Features

* allow `run_prog` to run without sub-parsers for single command programs ([f54ed2e](https://github.com/jayanthkoushik/shinyutils/commit/f54ed2ed2e3d259f6a1f623e06a36effb71c1f43))

### [12.2.1](https://github.com/jayanthkoushik/shinyutils/compare/v12.2.0...v12.2.1) (2021-12-09)


### Bug Fixes

* make `MatWrap` properties work for Python 3.8 and below ([79640a1](https://github.com/jayanthkoushik/shinyutils/commit/79640a169ab1e8ff9e02cf5f55ecc67cbdaae4ab))

## [12.2.0](https://github.com/jayanthkoushik/shinyutils/compare/v12.1.1...v12.2.0) (2021-12-08)


### Features

* add `add_short_full_helps` argument to `run_prog` ([bc5c85b](https://github.com/jayanthkoushik/shinyutils/commit/bc5c85b0ca3cedfa4438f730004d6782b12bc1d7))
* add `run_prog` function ([fdae9ae](https://github.com/jayanthkoushik/shinyutils/commit/fdae9aef2773c83c52837cc97977e4b7f8ba71c7))
* add more arguments to `run_prog` ([20cd2bd](https://github.com/jayanthkoushik/shinyutils/commit/20cd2bdf392d819d995492a17c5fb5881fe30aa0))
* allow specifying sub-commands with keyword arguments in `run_prog` ([745b4f2](https://github.com/jayanthkoushik/shinyutils/commit/745b4f28e4a9a633ddca42f63616e9ed4c69e369))


### Bug Fixes

* apply `add_short_full_helps` to subparsers in `run_prog` ([df8392b](https://github.com/jayanthkoushik/shinyutils/commit/df8392be5681aa11222ae428fc463711e5653932))

### [12.1.1](https://github.com/jayanthkoushik/shinyutils/compare/v12.1.0...v12.1.1) (2021-12-07)


### Bug Fixes

* use Python 3.7 compatible type annotations ([24d033c](https://github.com/jayanthkoushik/shinyutils/commit/24d033c6f579805d8a963066629f9b2ea8ea3d20))

## [12.1.0](https://github.com/jayanthkoushik/shinyutils/compare/v12.0.0...v12.1.0) (2021-12-06)


### Features

* use `corgy 4.1` to support Python 3.7 and 3.8 ([6b367df](https://github.com/jayanthkoushik/shinyutils/commit/6b367df4a49f0f90c45eb340d6ccf10f1688d1b8))


### Bug Fixes

* fix mock `trange` implementation ([4b5d4a2](https://github.com/jayanthkoushik/shinyutils/commit/4b5d4a20da5580703ff632bec8968c39410290ac))

## [12.0.0](https://github.com/jayanthkoushik/shinyutils/compare/v11.0.0...v12.0.0) (2021-12-05)


### ⚠ BREAKING CHANGES

* update `corgy` to version 4.0
* simplify `NNTrainer` var names
* refactor `pt` based on latest `corgy`

### Features

* call `MatWrap.configure` in `PlottingArgs.__init__` ([d7c1628](https://github.com/jayanthkoushik/shinyutils/commit/d7c1628c1c975cb86d455c5607e1163e54b8c728))
* mark wrapped names in `matwrap` as `Any` so `mypy` doesn't complain ([bf8c586](https://github.com/jayanthkoushik/shinyutils/commit/bf8c5866a7267f098e83c3a86818f6e32dc7956a))
* refactor `pt` based on latest `corgy` ([d9123e2](https://github.com/jayanthkoushik/shinyutils/commit/d9123e2c034ed323948e71558457449e54b2b74a))
* simplify `NNTrainer` var names ([7bf49d4](https://github.com/jayanthkoushik/shinyutils/commit/7bf49d4f7120867d0e2dd73c37e1ee76028436ea))


### Bug Fixes

* fix wrapped name import from `matwrap` ([ef4faae](https://github.com/jayanthkoushik/shinyutils/commit/ef4faae34bdf4b47ff732911f29ccb3b3b06f16f))


### build

* update `corgy` to version 4.0 ([d72b8d7](https://github.com/jayanthkoushik/shinyutils/commit/d72b8d78614dd52c142faaff4f6482d8f02afbf7))

## [11.0.0](https://github.com/jayanthkoushik/shinyutils/compare/v10.0.0...v11.0.0) (2021-12-02)


### ⚠ BREAKING CHANGES

* update to `corgy 3.1`

### Features

* add `py.typed` to inform type checkers about annotations ([2451394](https://github.com/jayanthkoushik/shinyutils/commit/245139443c38cac4747598e44dcafe5dd3d6ffa1))


### build

* update to `corgy 3.1` ([3a0526f](https://github.com/jayanthkoushik/shinyutils/commit/3a0526f96af9daffe3182d28389efef5ad3cdb51))

## [10.0.0](https://github.com/jayanthkoushik/shinyutils/compare/v9.3.0...v10.0.0) (2021-11-29)


### ⚠ BREAKING CHANGES

* update `corgy` to version 2.4
* refactor `pt` to work better with `Corgy`
* make `MatWrap.mpl/plt/sns` properties
* replace `MatWrap.add_plotting_args` with `PlottingArgs` class
* disalbe auto calling `conf_logging` and add parse arguments for it
* remove logng.build_log_argp
* make arguments to conf_logging keyword only
* rename 'color' dependency group to 'colors'
* remove plotting and pytorch dependency groups
* increase minimum Python version to 3.9

### Features

* add helps for argparse arguments ([a2ccde1](https://github.com/jayanthkoushik/shinyutils/commit/a2ccde1d1569dc918f11de40f3e624b12329a413))
* make `MatWrap.mpl/plt/sns` properties ([bae2f78](https://github.com/jayanthkoushik/shinyutils/commit/bae2f78c3f5b53fd2119041bade4f07f7f76a5df))
* make arguments to conf_logging keyword only ([3b194d6](https://github.com/jayanthkoushik/shinyutils/commit/3b194d60982af83890c6161b24f6accb37dbb68e))
* refactor `pt` to work better with `Corgy` ([869be1c](https://github.com/jayanthkoushik/shinyutils/commit/869be1cf41d23766c56ddbb842a1ca83fa767ee4))
* replace `MatWrap.add_plotting_args` with `PlottingArgs` class ([3974414](https://github.com/jayanthkoushik/shinyutils/commit/397441490c559b094982e317d5394c52b64ce18e))


* disalbe auto calling `conf_logging` and add parse arguments for it ([aaaecbf](https://github.com/jayanthkoushik/shinyutils/commit/aaaecbf5fda9a133b2c405d2f5107971332f8a34))
* remove logng.build_log_argp ([94ea6b9](https://github.com/jayanthkoushik/shinyutils/commit/94ea6b974dde5f94f50ea1090956a152349bce32))


### build

* increase minimum Python version to 3.9 ([3e16baf](https://github.com/jayanthkoushik/shinyutils/commit/3e16baf41a5b7098f3fd8af714a98a85699c4e66))
* remove plotting and pytorch dependency groups ([729e781](https://github.com/jayanthkoushik/shinyutils/commit/729e781163ab5346d144b449ee4013de79dc6469))

shinyutils/pt.py
"""Utilities for pytorch."""

try:
    import torch
except ImportError:
    raise ImportError("shinyutils.pt needs `pytorch`") from None

import inspect
import json
import warnings
from argparse import Action, ArgumentParser, ArgumentTypeError
from typing import (
    Annotated,
    Any,
    Callable,
    Iterable,
    Optional,
    overload,
    Sequence,
    Tuple,
    TYPE_CHECKING,
)
from unittest.mock import Mock

import torch.nn.functional as F
from corgy import Corgy, corgyparser
from corgy.types import KeyValuePairs, SubClass
from torch import nn
from torch.optim.lr_scheduler import _LRScheduler
from torch.optim.optimizer import Optimizer
from torch.utils.data import DataLoader, Dataset, TensorDataset

try:
    from tqdm import trange
except ImportError:
    warnings.warn("progress bar disabled: could not import `tqdm`", RuntimeWarning)

    class trange:  # type: ignore
        def __init__(self, n, *args, **kwargs):
            self._range = range(n)

        def __enter__(self):
            return self

        def __iter__(self):
            return iter(self._range)


if TYPE_CHECKING:
    import numpy as np

__all__ = ("DEFAULT_DEVICE", "PTOpt", "FCNet", "NNTrainer", "TBLogs")

DEFAULT_DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


class PTOpt(Corgy):
    """Wrapper around PyTorch optimizer and learning rate scheduler.

    Usage::

        >>> opt = PTOpt(Adam, {"lr": 0.001})
        >>> net = nn.Module(...)  # some network
        >>> opt.set_weights(net.parameters())
        >>> opt.zero_grad()
        >>> opt.step()
    """

    __slots__ = ("optimizer", "lr_scheduler")

    class _OptimizerSubClass(SubClass[Optimizer]):
        @classmethod
        def _choices(cls):
            return tuple(
                _c
                for _c in super()._choices()
                if _c.__module__ != "torch.optim._multi_tensor"
            )

    optim_cls: Annotated[
        _OptimizerSubClass, "optimizer sub class"
    ] = _OptimizerSubClass("Adam")

    optim_params: Annotated[
        KeyValuePairs, "arguments for the optimizer"
    ] = KeyValuePairs("")

    lr_sched_cls: Annotated[
        Optional[SubClass[_LRScheduler]], "learning rate scheduler sub class"
    ] = None

    lr_sched_params: Annotated[
        KeyValuePairs, "arguments for the learning rate scheduler"
    ] = KeyValuePairs("")

    @corgyparser("optim_params")
    @corgyparser("lr_sched_params")
    @staticmethod
    def _t_params(s: str) -> KeyValuePairs:
        dic = KeyValuePairs[str, str](s)
        for k, v in dic.items():
            v = json.loads(v)
            dic[k] = v
        return dic

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.optimizer = None
        self.lr_scheduler = None

    def set_weights(self, weights: Iterable[torch.Tensor]):
        """Set weights of underlying optimizer."""
        self.optimizer = self.optim_cls(weights, **self.optim_params)
        if self.lr_sched_cls is not None:
            self.lr_scheduler = self.lr_sched_cls(  # pylint: disable=not-callable
                self.optimizer, **self.lr_sched_params
            )

    @staticmethod
    def _better_lr_sched_repr(lr_sched: _LRScheduler) -> str:
        return (
            lr_sched.__class__.__name__
            + "(\n    "
            + "\n    ".join(
                f"{k}: {v}"
                for k, v in lr_sched.state_dict().items()
                if not k.startswith("_")
            )
            + "\n)"
        )

    def __repr__(self) -> str:
        if self.optimizer is None:
            return super().__repr__()
        r = repr(self.optimizer)
        if self.lr_scheduler is not None:
            r += f"\n{self._better_lr_sched_repr(self.lr_scheduler)}"
        return r

    def _ensure_initialized(self):
        if self.optimizer is None:
            raise TypeError("no weights set: call `PTOpt.set_weights` first")

    def zero_grad(self):
        """Call `zero_grad` on underlying optimizer."""
        self._ensure_initialized()
        self.optimizer.zero_grad()

    def step(self):
        """Call `step` on underlying optimizer, and lr scheduler (if present)."""
        self._ensure_initialized()
        self.optimizer.step()
        if self.lr_scheduler is not None:
            self.lr_scheduler.step()

    @staticmethod
    def add_help_args_to_parser(
        base_parser: ArgumentParser, group_title: Optional[str] = "pytorch help"
    ):
        """Add parser arguments for help on PyTorch optimizers and lr schedulers.

        Args:
            base_parser: `ArgumentParser` instance to add arguments to.
            group_title: Title of a new group to add arguments to. If `None`, arguments
                are added to the base parser instead. Default is `"pytorch help"`.

        Example::

            >>> arg_parser = ArgumentParser(
                    add_help=False, formatter_class=corgy.CorgyHelpFormatter
            )
            >>> PTOpt.add_help_args_to_parser(arg_parser)
            >>> arg_parser.print_help()
            pytorch help:
              --explain-optimizer cls  describe arguments of a torch optimizer
                                       (optional)
              --explain-lr-sched cls   describe arguments of a torch lr scheduler
                                       (optional)
            >>> arg_parser.parse_args(["--explain-optimizer", "Adamax"])
            Adamax(params, lr=0.002, betas=(0.9, 0.999), eps=1e-08, weight_decay=0)
            ...
        """

        class _ShowHelp(Action):
            def __call__(self, parser, namespace, values, option_string=None):
                cls_name = values.__name__
                cls_sig = inspect.signature(values)
                cls_doc = inspect.getdoc(values)
                print(f"{cls_name}{cls_sig}\n\n{cls_doc}")
                parser.exit()

        if group_title is not None:
            base_parser = base_parser.add_argument_group(group_title)  # type: ignore

        base_parser.add_argument(
            "--explain-optimizer",
            type=PTOpt._OptimizerSubClass,
            action=_ShowHelp,
            help="describe a pytorch optimizer",
            choices=PTOpt._OptimizerSubClass._choices(),
        )
        base_parser.add_argument(
            "--explain-lr-sched",
            type=SubClass[_LRScheduler],
            action=_ShowHelp,
            help="describe a pytorch lr scheduler",
            choices=SubClass[_LRScheduler]._choices(),
        )


class FCNet(Corgy, nn.Module):
    """Fully connected PyTorch network."""

    _ActType = Callable[..., torch.Tensor]
    _ActType.__metavar__ = "fun"  # type: ignore

    __slots__ = ("__dict__",)

    in_dim: Annotated[int, "number of input features"]
    out_dim: Annotated[int, "number of output features"]
    hidden_dims: Annotated[Sequence[int], "hidden layer dimensions"]
    hidden_act: Annotated[_ActType, "activation function for hidden layers"] = F.relu
    out_act: Annotated[
        Optional[_ActType], "activation function for output layer"
    ] = None

    @corgyparser("hidden_act")
    @corgyparser("out_act")
    @staticmethod
    def _activation_function(s: str) -> _ActType:
        try:
            return getattr(F, s)
        except AttributeError:
            raise ArgumentTypeError(
                f"`torch.nn.functional` has no attribute `{s}`"
            ) from None

    def __init__(self, **kwargs):
        nn.Module.__init__(self)
        Corgy.__init__(self, **kwargs)
        layer_sizes = [self.in_dim] + self.hidden_dims + [self.out_dim]
        self.layers = nn.ModuleList(
            [nn.Linear(ls, ls_n) for ls, ls_n in zip(layer_sizes, layer_sizes[1:])]
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """Forward a tensor through the network, and return the result.

        Args:
            x: Input tensor of shape `(batch_size, in_dim)`.
        """
        for layer in self.layers[:-1]:
            x = self.hidden_act(layer(x))
        x = self.layers[-1](x)
        if self.out_act is not None:
            x = self.out_act(x)  # pylint: disable=not-callable
        return x


class NNTrainer(Corgy):
    """Helper class for training a PyTorch model on a dataset."""

    __slots__ = ("_dataset", "_data_loader")

    iters: Annotated[int, "number of training iterations"]
    ptopt: Annotated[PTOpt, "optimizer and learning rate scheduler"]
    batch_size: Annotated[int, "batch size for training"] = 8
    data_workers: Annotated[int, "number of workers for loading data"] = 0
    shuffle_data: Annotated[bool, "whether to shuffle the dataset"] = True
    pin_cuda: Annotated[bool, "whether to pin data to CUDA memory"] = True
    drop_last: Annotated[bool, "whether to drop the last incomplete batch"] = False
    pbar_desc: Annotated[str, "description for training progress bar"] = "Training"

    @overload
    def set_dataset(self, value: Dataset):
        ...

    @overload
    def set_dataset(self, value: Tuple[torch.Tensor, ...]):
        ...

    @overload
    def set_dataset(self, value: Tuple["np.ndarray", ...]):
        ...

    def set_dataset(self, value):
        """Set the training data.

        Args:
            value: `torch.utils.data.Dataset` instance, or tuple of `torch.Tensor` or
                `np.ndarray` objects.
        """
        if isinstance(value, Dataset):
            self._dataset = value
        elif isinstance(value, tuple):
            if not isinstance(value[0], torch.Tensor):
                value = [torch.from_numpy(val_i) for val_i in value]
            self._dataset = TensorDataset(*value)
        else:
            raise ValueError(f"can't set dataset from type `{type(value)}`")

        self._data_loader = DataLoader(
            self._dataset,
            batch_size=self.batch_size,
            num_workers=self.data_workers,
            shuffle=self.shuffle_data,
            pin_memory=self.pin_cuda,
            drop_last=self.drop_last,
        )

    def train(
        self,
        model: nn.Module,
        loss_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
        post_iter_hook: Optional[
            Callable[
                [int, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, Any], None
            ]
        ] = None,
    ):
        """Train a model.

        Args:
            model: Model (`nn.Module` instance) to train.
            loss_fn: Loss function mapping input tensors to a loss tensor.
            post_iter_hook: Optional callback function to call after each iteration.
                The function will be called with arguments
                `(iteration, x_batch, y_batch, yhat_batch, loss, pbar)`.
        """
        if self._dataset is None:
            raise RuntimeError("dataset not set: call `set_dataset` before `train`")
        bat_iter = iter(self._data_loader)

        model = model.to(DEFAULT_DEVICE)
        self.ptopt.set_weights(model.parameters())

        for _iter in (pbar := trange(self.iters, desc=self.pbar_desc)):
            try:
                x_bat, y_bat = next(bat_iter)
            except StopIteration:
                bat_iter = iter(self._data_loader)
                x_bat, y_bat = next(bat_iter)
            x_bat, y_bat = x_bat.to(DEFAULT_DEVICE), y_bat.to(DEFAULT_DEVICE)

            yhat_bat = model(x_bat)
            loss = loss_fn(yhat_bat, y_bat)
            pbar.set_postfix(loss=float(loss))

            self.ptopt.zero_grad()
            loss.backward()
            self.ptopt.step()

            if post_iter_hook is not None:
                post_iter_hook(_iter, x_bat, y_bat, yhat_bat, loss, pbar)


class TBLogs:
    """TensorBoard logs type.

    Args:
        path: Path to log directory. If `None` (default), a mock instance is
            returned.

    Usage::

        tb_logs = TBLogs("tmp/tb")
        tb_logs.writer  # `SummaryWriter` instance
        TBLogs.mock  # mock instance
    """

    __metavar__ = "dir"
    _mock: Optional["TBLogs"] = None

    def __init__(self, path: Optional[str] = None):
        if path is not None:
            try:
                from torch.utils.tensorboard import SummaryWriter
            except ImportError:
                raise RuntimeError("tensorboard not installed") from None
            self.writer = SummaryWriter(path)
        else:
            self.writer = Mock()

    @classmethod
    @property
    def mock(cls) -> "TBLogs":
        """Mock instace that no-ops for every call."""
        if cls._mock is None:
            cls._mock = cls()
        return cls._mock

    def __repr__(self) -> str:
        if isinstance(self.writer, Mock):
            return "TBLogs.mock"
        return f"TBLogs({self.writer.log_dir!r})"
